Add auto-backup

leonklingele opened the pull request:
Implements #62

leonklingele commented:
Again — should have created a separate PR for creating constants for the ConfigType

bndw commented:
Looks like the move to the `src` folder broke this one. Once it's rebased I'll take a look

leonklingele commented:
Rebased.

bndw commented:
This implementation does not include the `max_backups` setting discussed in https://github.com/bndw/pick/issues/62. What are your thoughts on including that in this PR?

leonklingele commented:
It's not easy. We would either need to parse all backup file names and sort them by date, or simply sort by `os.FileInfo.ModTime()`. I don't like both ways.

Until I figured out a better way, this PR should be merged ASAP.
I will work on the `max_backup` feature this weekend.

leonklingele commented:
Implemented `max_backups`.

leonklingele commented on config.toml.in:
```diff
@@ -7,6 +7,8 @@
 # The storage section controls the configuration of the backend.
 [storage]
 type = "file"
+auto_backup = true
```
Description of these two new directives is missing.

leonklingele replied:
Should we move `auto_backup` and `max_backups` to a new `[storage.backup]` / `[general.backup]` section?

bndw replied:
I suppose a "backup" is a concept of the safe, so it makes sense to set `auto_backup` outside of the `[storage]` table.

I could image setting this at the top level, or in a `safe` or `general` table. I'm leaning toward top level.

bndw replied:
Looks like you've added ` general` section [here] (https://github.com/leonklingele/pick/blob/61bc425e552cb72940d581a4eae9cc99254c1f10/config.toml.in#L12), so let's stick `auto_backup` in there, too.

leonklingele replied:
Hmm, we have a `auto_backup` and `max_backups`.
Now imagine we decide to add a new `backup_format` to configure the backup-date format (or yet another backup directive). In my opinion it would make sense to group them in a more specific section like `[general.backup]`.
What do you think?

```toml
[general]
auto_backup = true
max_backups = 100
```

vs 

```toml
[general.backup]
auto = true
max = 100
```

bndw replied:
In that case, why not use a dedicated backup table?

```toml
[backup]
auto = true
max = 100
```

leonklingele replied:
👍

leonklingele commented:
Updated, rebased & merge target changed.

I decided to put the backup configuration to `[storage.backup]` as it's clearly related to `storage`.
Also it's not as simple to move it to the main config (`[backup]`)

leonklingele commented on src/backends/disk.go:
```diff
@@ -74,8 +83,59 @@ func (db *DiskBackend) Save(data []byte) error {
+		return &errors.BackupDisabled{}
+	} else if db.backupConfig.MaxFiles > 0 {
+		// Subtract one as we are about to create another backup
+		db.cleanOldBackups(db.backupConfig.MaxFiles - 1)
```
Why? We don't really care if there were issues with removing the old backups?!

bndw replied:
It's unexpected behavior, and as a user that is not something I want from my password manager.

If the error is not fatal, which I agree it is not, we should at least surface something to the user , simlilar to https://github.com/bndw/pick/pull/68#discussion-diff-79317027R114, since there are multiple operations performed.

leonklingele replied:
Why should a user care if his old backups were removed without success?

leonklingele replied:
Nvm, error is now logged.

leonklingele commented on src/backends/disk.go:
```diff
@@ -74,8 +83,59 @@ func (db *DiskBackend) Save(data []byte) error {
+	for _, f := range filesSorted[:len(filesSorted)-max] {
+		p := fmt.Sprintf("%s/%s", db.backupConfig.DirPath, f.Name())
+		if err := os.Remove(p); err != nil {
+			fmt.Println("Error", err.Error())
```
Hmm. We do not `return err` here as deletion of the other backup files still could succeed.
Returning just the last error also doesn't make sense. That's why I decided to not return an error at all.

bndw replied:
Your point about just returning the last error is valid and we are printing the error, if one is encountered.

I'd feel better if the printed message provided a bit more context, like a `Error removing backup` prefix or something.

leonklingele replied:
Updated error message.

bndw commented:
LGTM
